Adds support for pypoetry (https://python-poetry.org)

Trybnetic opened the pull request:
This PR introduces the Python packaging and dependency management [poetry](https://python-poetry.org) to this project.

Most stuff is handled in the `pyproject.toml` which is also makes other files (e.g. setup.py, requirements.txt, MANIFEST.in) unnecessary and helps to keep the project cleaner. Further, more and more third-party applications like pytest, pylint, etc are supported to be configured in the toml file. 

To test out, install the package with 
```
poetry install
```

and run the tests with 
```
poetry run pytest
```

This PR is yet, just a draft, as there can (and in my opinion) should be more things cleaned up in the process. This includes:

- [x] check what other configs can be transferred to the pyproject.toml
- [x] Is there any benefit from using tox any longer?
- [x] If we want to keep tox adjust the tox.ini file accordingly, if not adjust the Makefile
- [x] Adapt .travis.yml or setup Github Actions accordingly

Trybnetic commented:
Currently, the docs are not building due to the poetry installation. This is due to how we have set up `conf.py`. This should definitely be fixed before merging this PR. I will try to do this asap.

Trybnetic commented:
It is also questionable whether it is worth to keep the `setup.cfg` file. The pylint conf is already included in the `pyproject.toml` and the pycodestyle config is only about the max-length and will be hopefully supported to be configured in the `pyproject.toml` soon.

Trybnetic commented:
I think this PR is now almost ready to be merged. The only things that should be decided and done beforehand are

- [ ] Either fix `tox.ini` or remove `tox.ini` and `Makefile`
- [ ] Remove `setup.cfg` as I think it is useless now (see comment above)

derNarr reviewed (changes requested):
This PR looks really good as well. Thanks @Trybnetic !

It is good to see of how many config files we can get rid of. In this respect poetry seems to keep its promise.

`peotry run pylint` does not seem to work. It might need to be replaced with `peotry run pylint pyndl`.

Running `poetry run sphinx-build -b html doc/source doc/build/html` results in an exception:
```
  File "pyndl/doc/source/conf.py", line 23, in <module>
    author =  pkg.__author__
AttributeError: module 'pyndl' has no attribute '__author__'
```

With `poetry publish` we will loose the ability to gpg sign the package. This seems not to be so much of a problem and is a wontfix at the moment in peotry. The whole security of PyPI seems to rely on a proper https connection and that PyPI is not compromised.
https://github.com/python-poetry/poetry/issues/127

derNarr commented:
It would be nice to include the doctesting automagically into the invocation of `poetry run pytest`

derNarr commented:
General plan would be to first merge PR #224 with version number `0.9.0` and then publish version `1.0.0` while merging this PR. What do your think? With issue #208 and #223 and #217, #218, #219 we then can bump version to `1.0.X`.

Poetry seems to build a wheel as well, which would fix #212. And when we don't sign the packages anymore, we might use some direct publishing pipeline with github actions as well.

derNarr commented on pyproject.toml:
```diff
@@ -0,0 +1,73 @@
+build = "build.py"
+
+[tool.poetry.dependencies]
+python = ">=3.8,<3.10"  # Compatible python versions must be declared here
```
Does anything breaks for Python 3.10? Or why are we restricting python to below 3.10?

Trybnetic commented:
> `peotry run pylint` does not seem to work. It might need to be replaced with `peotry run pylint pyndl`.

You are completely right on that. Fixed it in the documentation

> Running `poetry run sphinx-build -b html doc/source doc/build/html` results in an exception:
> 
> ```
>   File "pyndl/doc/source/conf.py", line 23, in <module>
>     author =  pkg.__author__
> AttributeError: module 'pyndl' has no attribute '__author__'
> ```

Nice notice, this has been caused by 2a21bc4ae9d3d872797f8e82174a43d0dca26828 as I removed a bit too much information. Fixed that now by extracting the same info using importlib. However, this means the author information is no longer available as `pyndl.__author__`, however, this does not seem to be widely used anyway..

> It would be nice to include the doctesting automagically into the invocation of poetry run pytest

Doctests are now included when running pytest. I hesitated doing that as I don't know with this config how to only test the unit tests

For me all the documentation issues seem not too much connected to this branch and I would rather resolve them a separate Pull Request (#223 was intended for that even though I merged its current status already into this PR)

derNarr commented:
Creating the documentation offline with `pypoetry` works now. Building a sdist and wheel works as well. I have not tested publishing yet.
Doctests fail on macos, but we should make `threading` the default method anyways see #184 .

derNarr commented:
Ok, we should simply add `method=threading` into the documentation for now as `threading` is not available for the `wh` code right now.

Trybnetic commented on doc/source/examples.rst:
```diff
@@ -105,8 +105,8 @@ pyndl.ndl.ndl
-former one using `openMP <http://www.openmp.org/>`_ and therefore being expected
-to be much faster when analyzing larger data. Besides, you can set three
+former one using `openMP <http://www.openmp.org/>`_ and therefore being
+expected to be faster when analyzing larger data. Besides, you can set three
```
We should add a sentence here that OpenMP is Linux only and other OS should use threading to make clear why threading here is used instead of openmp

derNarr replied:
Something along the lines:
> Unfortunately, ``method=openmp`` is only available on Linux right now, therefore all examples use ``method='threading'``.
Can you add this into the text? Thanks!

Trybnetic commented:
To sum up the changes proposed in this PR:

- [x] introduces python poetry to maintain and build the package
- [x] allows building wheels for PyPI  (fixes #212) 
- [x] uses `pyproject.toml` instead of `setup.py` and thus fixes  #143 and fixes #137 
- [x] sets up Github Actions for general testing (replaces Travis)

As said earlier, I would leave the https://github.com/quantling/pyndl/labels/documentation related issues for a separate Pull Request as most of them are unrelated to the changes proposed here and require some substantial effort on their own.

derNarr reviewed (dismissed):
Looks good now.
